{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63097c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "081ea136",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "800\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>snp</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>score</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>S</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>155.0</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>Node2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>R</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>26.0</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>56.0</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site            gene direction   snp dimer  apobec  aa_position parent  \\\n",
       "0   3111  OPG003_CDS_176   reverse  G->A    GA    True          2.0  Node1   \n",
       "1  39139  OPG057_CDS_142   reverse  C->T    TC    True          3.0  Node1   \n",
       "2  73239  OPG093_CDS_107   forward  G->A    GA    True          2.0  Node1   \n",
       "3  74205  OPG094_CDS_106   forward  G->A    GA    True          1.0  Node1   \n",
       "4  77383  OPG098_CDS_102   forward  G->A    GA    True          2.0  Node1   \n",
       "\n",
       "  parent_codon parent_aa  child child_codon child_aa mutation_category  score  \\\n",
       "0          TCC         S  Node2         TTC        F     nonsynonymous  155.0   \n",
       "1          GCG         A  Node2         GCA        A        synonymous    NaN   \n",
       "2          AGA         R  Node2         AAA        K     nonsynonymous   26.0   \n",
       "3          GAA         E  Node2         AAA        K     nonsynonymous   56.0   \n",
       "4          TGA         *  Node2         TAA        *        synonymous    NaN   \n",
       "\n",
       "                prediction  homoplasy  occurrence  \n",
       "0                  radical      False           1  \n",
       "1                      NaN      False           1  \n",
       "2             conservative      False           1  \n",
       "3  moderately conservative      False           1  \n",
       "4                      NaN      False           1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_pos_file = \"data/B.1_2022-08-22.og.aln.pruned.tree.amino_acid.reconstruction.csv\"\n",
    "article_df = pd.read_csv(article_pos_file, sep=',')\n",
    "print(len(article_df))\n",
    "article_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c616a053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['site', 'gene', 'direction', 'snp', 'dimer', 'apobec', 'aa_position',\n",
       "       'parent', 'parent_codon', 'parent_aa', 'child', 'child_codon',\n",
       "       'child_aa', 'mutation_category', 'score', 'prediction', 'homoplasy',\n",
       "       'occurrence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e93d2114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>score</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>S</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>155.0</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>Node2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>R</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>26.0</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>56.0</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site            gene direction REF ALT dimer  apobec  aa_position parent  \\\n",
       "0   3111  OPG003_CDS_176   reverse   G   A    GA    True          2.0  Node1   \n",
       "1  39139  OPG057_CDS_142   reverse   C   T    TC    True          3.0  Node1   \n",
       "2  73239  OPG093_CDS_107   forward   G   A    GA    True          2.0  Node1   \n",
       "3  74205  OPG094_CDS_106   forward   G   A    GA    True          1.0  Node1   \n",
       "4  77383  OPG098_CDS_102   forward   G   A    GA    True          2.0  Node1   \n",
       "\n",
       "  parent_codon parent_aa  child child_codon child_aa mutation_category  score  \\\n",
       "0          TCC         S  Node2         TTC        F     nonsynonymous  155.0   \n",
       "1          GCG         A  Node2         GCA        A        synonymous    NaN   \n",
       "2          AGA         R  Node2         AAA        K     nonsynonymous   26.0   \n",
       "3          GAA         E  Node2         AAA        K     nonsynonymous   56.0   \n",
       "4          TGA         *  Node2         TAA        *        synonymous    NaN   \n",
       "\n",
       "                prediction  homoplasy  occurrence  \n",
       "0                  radical      False           1  \n",
       "1                      NaN      False           1  \n",
       "2             conservative      False           1  \n",
       "3  moderately conservative      False           1  \n",
       "4                      NaN      False           1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df['REF'] = article_df.apply(lambda row: row['snp'].split('->')[0], axis = 1)\n",
    "article_df['ALT'] = article_df.apply(lambda row: row['snp'].split('->')[1], axis = 1)\n",
    "article_df = article_df[['site', 'gene', 'direction', 'REF', \"ALT\", 'dimer', 'apobec', 'aa_position',\n",
    "       'parent', 'parent_codon', 'parent_aa', 'child', 'child_codon',\n",
    "       'child_aa', 'mutation_category', 'score', 'prediction', 'homoplasy',\n",
    "       'occurrence']]\n",
    "article_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7265889",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genome(fasta_file):\n",
    "    \"\"\"Load the genome from a FASTA file.\"\"\"\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    new_genome = genome[\"NC_063383.1\"].seq\n",
    "    return new_genome\n",
    "\n",
    "def load_gtf(gtf_file):\n",
    "    \"\"\"Load the GTF file into a pandas DataFrame.\"\"\"\n",
    "    gtf_columns = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "    gtf = pd.read_csv(gtf_file, sep=\"\\t\", comment='#', header=None, names=gtf_columns)\n",
    "    return gtf\n",
    "\n",
    "def find_parent_mutation(genome, position, parent, table=article_df):\n",
    "    parent_filter_df = table[table['child']==parent]\n",
    "    parent_filter_df = parent_filter_df.reset_index(drop=True)\n",
    "    new_parent = parent_filter_df.loc[0, \"parent\"]\n",
    "    sites_list = parent_filter_df.site.tolist()\n",
    "    if position in sites_list or new_parent == \"Node1\":\n",
    "        if new_parent == \"Node1\":\n",
    "            return genome\n",
    "        else:\n",
    "            new_df = parent_filter_df[parent_filter_df['site']==position]\n",
    "            new_df = new_df.reset_index(drop=True)\n",
    "            alt_nucl = new_df.loc[0, \"ALT\"]\n",
    "            new_genome = list(genome)\n",
    "            new_genome[position-1] = alt_nucl\n",
    "            return Seq(''.join(new_genome))\n",
    "    else:\n",
    "        return find_parent_mutation(genome, position, new_parent)\n",
    "\n",
    "def get_cds(gtf, chromosome, position):\n",
    "    \"\"\"Find the CDS where the mutation is located.\"\"\"\n",
    "    cds_rows = gtf[(gtf['seqname'] == chromosome) & \n",
    "                   (gtf['feature'] == 'CDS') & \n",
    "                   (gtf['start'] <= position) & \n",
    "                   (gtf['end'] >= position)]\n",
    "    return cds_rows\n",
    "\n",
    "def translate_codon(codon):\n",
    "    \"\"\"Translate a codon to its corresponding amino acid.\"\"\"\n",
    "    return str(Seq(codon).translate(table=CodonTable.unambiguous_dna_by_id[11]))\n",
    "\n",
    "def detect_mutation_impact(genome, gtf, chromosome, position, ref, alt):\n",
    "    \"\"\"Detect the amino acid change caused by a mutation.\"\"\"\n",
    "    cds = get_cds(gtf, chromosome, position)\n",
    "    \n",
    "    if cds.empty:\n",
    "        ref_nucl = genome[position-1]\n",
    "        return '-', '-', '-', '-', ref_nucl, '-'\n",
    "    \n",
    "    strand = cds.iloc[0]['strand']\n",
    "    gene_start = cds.iloc[0]['start']\n",
    "    \n",
    "    # Calculate the codon position\n",
    "    codon_start = gene_start + ((position - gene_start) // 3) * 3\n",
    "    \n",
    "    # Extract the codon\n",
    "    codon_seq = genome[codon_start-1:codon_start+2]\n",
    "    if strand == '-':\n",
    "        codon_seq = codon_seq.reverse_complement()\n",
    "        alt = str(Seq(alt).reverse_complement())\n",
    "    \n",
    "    # Translate the original codon\n",
    "    original_amino_acid = translate_codon(codon_seq)\n",
    "    \n",
    "    # Create the mutated codon\n",
    "    mutation_index = (position - codon_start) % 3\n",
    "    if strand == '-':\n",
    "        if mutation_index == 0:\n",
    "            mutation_index = 2\n",
    "        elif mutation_index == 2:\n",
    "            mutation_index = 0\n",
    "\n",
    "    mutated_codon = list(codon_seq)\n",
    "    mutated_codon[mutation_index] = alt\n",
    "    mutated_codon = Seq(''.join(mutated_codon))\n",
    "    \n",
    "    # Translate the mutated codon\n",
    "    mutated_amino_acid = translate_codon(mutated_codon)\n",
    "    \n",
    "    ref_nucl = genome[position-1]\n",
    "    pos = mutation_index+1\n",
    "    \n",
    "    return original_amino_acid, mutated_amino_acid, codon_seq, mutated_codon, ref_nucl, pos\n",
    "\n",
    "\n",
    "\n",
    "# BODY\n",
    "genome = load_genome(os.path.join(r\"C:\\Users\\User\\Documents\\msu\\diplom\\APOBEC_mutagenesis\\MPXV\", \"NC_063383.1.fasta\"))\n",
    "gtf = load_gtf(os.path.join(r\"C:\\Users\\User\\Documents\\msu\\diplom\\APOBEC_mutagenesis\\MPXV\\aa_APOBEC_pos\", \"GCF_014621545.1_ASM1462154v1_genomic.220824.gff\"))\n",
    "chromosome = \"NC_063383.1\"\n",
    "\n",
    "original_aa_list = []\n",
    "mutated_aa_list = []\n",
    "original_codon_list = []\n",
    "mutated_codon_list = []\n",
    "ref_nucl_list = []\n",
    "nucl_pos_list = []\n",
    "mutation_category = []\n",
    "\n",
    "##duplicated data\n",
    "duplicated_df_science = article_df[article_df.duplicated('site', keep=False) == True]\n",
    "duplicated_df_science = duplicated_df_science.reset_index(drop=True)\n",
    "duplicates_list = duplicated_df_science.site.unique()\n",
    "\n",
    "for row in range(0, len(article_df)):\n",
    "    position = article_df.loc[row, \"site\"] \n",
    "    ref_nucl = article_df.loc[row, \"REF\"]\n",
    "    alt_nucl = article_df.loc[row, \"ALT\"]\n",
    "\n",
    "    if position in duplicates_list:\n",
    "        node = article_df.loc[row, \"parent\"]\n",
    "        new_genome = find_parent_mutation(genome, position, node)\n",
    "        original_aa, mutated_aa, original_codon, mutated_codon, ref, nucl_pos = detect_mutation_impact(new_genome, gtf, chromosome, position, ref_nucl, alt_nucl)\n",
    "    else:\n",
    "        original_aa, mutated_aa, original_codon, mutated_codon, ref, nucl_pos = detect_mutation_impact(genome, gtf, chromosome, position, ref_nucl, alt_nucl)\n",
    "        \n",
    "    # CDS empty\n",
    "    if original_aa == \"-\":\n",
    "        mutation_category.append(\"intergenic\")\n",
    "        original_aa_list.append('-')\n",
    "        mutated_aa_list.append('-')\n",
    "        original_codon_list.append('-')\n",
    "        mutated_codon_list.append('-')\n",
    "        nucl_pos_list.append('-')\n",
    "        ref_nucl_list.append(ref)\n",
    "    else:\n",
    "        original_aa_list.append(original_aa)\n",
    "        mutated_aa_list.append(mutated_aa)\n",
    "        original_codon_list.append(str(original_codon))\n",
    "        mutated_codon_list.append(str(mutated_codon))\n",
    "        ref_nucl_list.append(ref)\n",
    "        nucl_pos_list.append(nucl_pos)\n",
    "\n",
    "        if mutated_aa == original_aa:\n",
    "            mutation_category.append(\"synonymous\")\n",
    "        elif mutated_aa == \"*\":\n",
    "            mutation_category.append(\"nonsense\")\n",
    "        else:\n",
    "            mutation_category.append(\"nonsynonymous\")\n",
    "            \n",
    "article_df[\"REF_v2\"] = ref_nucl_list\n",
    "article_df[\"nucl_pos_codon\"] = nucl_pos_list\n",
    "article_df[\"parent_codon_v2\"] = original_codon_list\n",
    "article_df[\"parent_aa_v2\"] = original_aa_list\n",
    "article_df[\"child_codon_v2\"] = mutated_codon_list\n",
    "article_df[\"mutated_aa_v2\"] = mutated_aa_list\n",
    "article_df[\"mutation_category_v2\"] = mutation_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "28f094c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon_v2</th>\n",
       "      <th>parent_aa_v2</th>\n",
       "      <th>child_codon_v2</th>\n",
       "      <th>mutated_aa_v2</th>\n",
       "      <th>mutation_category_v2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>...</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>ATC</td>\n",
       "      <td>I</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "      <td>AAG</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>...</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>...</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "      <td>AAG</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>127220</td>\n",
       "      <td>OPG148_CDS_52</td>\n",
       "      <td>forward</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node2</td>\n",
       "      <td>CTC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>144527</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Node2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>79348</td>\n",
       "      <td>OPG102_CDS_98</td>\n",
       "      <td>forward</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>...</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>ATC</td>\n",
       "      <td>I</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>86979</td>\n",
       "      <td>OPG108_CDS_92</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GC</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGC</td>\n",
       "      <td>...</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>133714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Node1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>800 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site            gene direction REF ALT dimer  apobec  aa_position  \\\n",
       "0      3111  OPG003_CDS_176   reverse   G   A    GA    True          2.0   \n",
       "1     39139  OPG057_CDS_142   reverse   C   T    TC    True          3.0   \n",
       "2     73239  OPG093_CDS_107   forward   G   A    GA    True          2.0   \n",
       "3     74205  OPG094_CDS_106   forward   G   A    GA    True          1.0   \n",
       "4     77383  OPG098_CDS_102   forward   G   A    GA    True          2.0   \n",
       "..      ...             ...       ...  ..  ..   ...     ...          ...   \n",
       "795  127220   OPG148_CDS_52   forward   C   T    TC    True          3.0   \n",
       "796  144527             NaN       NaN   C   T    TC    True          NaN   \n",
       "797   79348   OPG102_CDS_98   forward   C   T    TC    True          1.0   \n",
       "798   86979   OPG108_CDS_92   reverse   C   T    GC   False          2.0   \n",
       "799  133714             NaN       NaN   C   T    TC    True          NaN   \n",
       "\n",
       "    parent parent_codon  ...               prediction homoplasy occurrence  \\\n",
       "0    Node1          TCC  ...                  radical     False          1   \n",
       "1    Node1          GCG  ...                      NaN     False          1   \n",
       "2    Node1          AGA  ...             conservative     False          1   \n",
       "3    Node1          GAA  ...  moderately conservative     False          1   \n",
       "4    Node1          TGA  ...                      NaN     False          1   \n",
       "..     ...          ...  ...                      ...       ...        ...   \n",
       "795  Node2          CTC  ...                      NaN     False          1   \n",
       "796  Node2          NaN  ...                      NaN     False          1   \n",
       "797  Node1          CAT  ...  moderately conservative     False          1   \n",
       "798  Node1          AGC  ...             conservative     False          1   \n",
       "799  Node1          NaN  ...                      NaN     False          1   \n",
       "\n",
       "    REF_v2 nucl_pos_codon  parent_codon_v2 parent_aa_v2  child_codon_v2  \\\n",
       "0        G              3              ATC            I             ATT   \n",
       "1        C              1              GAG            E             AAG   \n",
       "2        G              1              GAT            D             AAT   \n",
       "3        G              3              ATG            M             ATA   \n",
       "4        G              1              GAG            E             AAG   \n",
       "..     ...            ...              ...          ...             ...   \n",
       "795      C              2              TCA            S             TTA   \n",
       "796      C              1              CCT            P             TCT   \n",
       "797      C              3              ATC            I             ATT   \n",
       "798      C              3              GAG            E             GAA   \n",
       "799      C              -                -            -               -   \n",
       "\n",
       "     mutated_aa_v2 mutation_category_v2  \n",
       "0                I           synonymous  \n",
       "1                K        nonsynonymous  \n",
       "2                N        nonsynonymous  \n",
       "3                I        nonsynonymous  \n",
       "4                K        nonsynonymous  \n",
       "..             ...                  ...  \n",
       "795              L        nonsynonymous  \n",
       "796              S        nonsynonymous  \n",
       "797              I           synonymous  \n",
       "798              E           synonymous  \n",
       "799              -           intergenic  \n",
       "\n",
       "[800 rows x 26 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b2ce08e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "does not match to reference:  2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon_v2</th>\n",
       "      <th>parent_aa_v2</th>\n",
       "      <th>child_codon_v2</th>\n",
       "      <th>mutated_aa_v2</th>\n",
       "      <th>mutation_category_v2</th>\n",
       "      <th>new_col</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>168120</td>\n",
       "      <td>OPG193_CDS_17</td>\n",
       "      <td>forward</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node386</td>\n",
       "      <td>TTT</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>81977</td>\n",
       "      <td>OPG105_CDS_95</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node561</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site           gene direction REF ALT dimer  apobec  aa_position  \\\n",
       "268  168120  OPG193_CDS_17   forward   T   C   NaN   False          2.0   \n",
       "423   81977  OPG105_CDS_95   forward   G   A    GA    True          1.0   \n",
       "\n",
       "      parent parent_codon  ... homoplasy occurrence REF_v2 nucl_pos_codon  \\\n",
       "268  Node386          TTT  ...     False          1      C              1   \n",
       "423  Node561          GAC  ...     False          1      A              3   \n",
       "\n",
       "    parent_codon_v2  parent_aa_v2 child_codon_v2  mutated_aa_v2  \\\n",
       "268             CTT             L            CTT              L   \n",
       "423             CAA             Q            CAA              Q   \n",
       "\n",
       "     mutation_category_v2 new_col  \n",
       "268            synonymous       -  \n",
       "423            synonymous       -  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_col = []\n",
    "for row in range(0, len(article_df)):\n",
    "    if article_df.loc[row, \"REF\"] != article_df.loc[row, \"REF_v2\"]:\n",
    "        new_col.append('-')\n",
    "    else:\n",
    "        new_col.append(\"+\")\n",
    "article_df[\"new_col\"] = new_col\n",
    "print('does not match to reference: ', len(article_df[article_df['new_col']=='-']))\n",
    "article_df[article_df['new_col']=='-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "97e0b6e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "724"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter only APOBEC mutations\n",
    "df1_new1 = article_df[(article_df['REF_v2']==\"G\") & (article_df['ALT']==\"A\")]\n",
    "df1_new2 = article_df[(article_df['REF_v2']==\"C\") & (article_df['ALT']==\"T\")]\n",
    "df1_new = pd.concat([df1_new1, df1_new2])\n",
    "df1_new = df1_new.reset_index(drop=True)\n",
    "len(df1_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1897770b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "641"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter TC and GA motifs\n",
    "from Bio import SeqIO\n",
    "\n",
    "def load_genome(fasta_file):\n",
    "    \"\"\"Load the genome from a FASTA file.\"\"\"\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    new_genome = genome[\"NC_063383.1\"].seq\n",
    "    return new_genome\n",
    "\n",
    "def find_parent_mutations(genome, parent, table=article_df):\n",
    "    parent_filter_df = table[table['child']==parent]\n",
    "    parent_filter_df = parent_filter_df.reset_index(drop=True)\n",
    "    new_parent = parent_filter_df.loc[0, \"parent\"]\n",
    "    if new_parent == \"Node1\":\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[0, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return Seq(''.join(new_genome))\n",
    "    else:\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[0, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return find_parent_mutations(new_genome, new_parent)\n",
    "    \n",
    "    \n",
    "genome = load_genome(os.path.join(r\"C:\\Users\\User\\Documents\\msu\\diplom\\APOBEC_mutagenesis\\MPXV\", \"NC_063383.1.fasta\"))\n",
    "motifs = []\n",
    "for row in range(0, len(df1_new)):\n",
    "    pos = df1_new.loc[row, 'site']\n",
    "    ref = df1_new.loc[row, 'REF']\n",
    "    node = df1_new.loc[row, 'parent']\n",
    "    if node != \"Node1\":\n",
    "        new_genome = find_parent_mutations(genome, node)\n",
    "        if ref == \"C\":\n",
    "            motifs.append(new_genome[pos-2:pos])\n",
    "        else:\n",
    "            motifs.append(new_genome[pos-1:pos+1])\n",
    "    else:\n",
    "        if ref == \"C\":\n",
    "            motifs.append(genome[pos-2:pos])\n",
    "        else:\n",
    "            motifs.append(genome[pos-1:pos+1])\n",
    "\n",
    "df1_new['motif'] = motifs\n",
    "\n",
    "df1_new1 = df1_new[df1_new['motif'] == 'GA']\n",
    "df1_new2 = df1_new[df1_new['motif'] == 'TC']\n",
    "df1_new = pd.concat([df1_new1, df1_new2])\n",
    "len(df1_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5565595c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "631"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter Node1\n",
    "df1_new = df1_new[df1_new['parent'] != 'Node1']\n",
    "df1_new = df1_new.reset_index(drop=True)\n",
    "len(df1_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9539ae71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        #CHROM  position REF ALT nucl_pos_codon parent_codon parent_aa  \\\n",
      "0  NC_063383.1        23   C   T              -            -         -   \n",
      "1  NC_063383.1        60   C   T              -            -         -   \n",
      "2  NC_063383.1        78   C   T              -            -         -   \n",
      "3  NC_063383.1        82   C   T              -            -         -   \n",
      "4  NC_063383.1        98   C   T              -            -         -   \n",
      "\n",
      "  child_codon mutated_aa mutation_category  \n",
      "0           -          -        intergenic  \n",
      "1           -          -        intergenic  \n",
      "2           -          -        intergenic  \n",
      "3           -          -        intergenic  \n",
      "4           -          -        intergenic  \n"
     ]
    }
   ],
   "source": [
    "df2 = pd.read_csv(\"APOBEC_targets_aa.csv\", sep='\\t')\n",
    "print(df2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "12a48591",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>site</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>ALT</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon_v2</th>\n",
       "      <th>parent_aa_v2</th>\n",
       "      <th>child_codon_v2</th>\n",
       "      <th>mutated_aa_v2</th>\n",
       "      <th>mutation_category_v2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>186165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM    site REF_v2 ALT nucl_pos_codon parent_codon_v2 parent_aa_v2  \\\n",
       "0  NC_063383.1  186165      G   A              1             GAT            D   \n",
       "1  NC_063383.1   55133      G   A              3             CTC            L   \n",
       "2  NC_063383.1   55133      G   A              3             CTC            L   \n",
       "3  NC_063383.1   55133      G   A              3             CTC            L   \n",
       "4  NC_063383.1   55133      G   A              3             CTC            L   \n",
       "\n",
       "  child_codon_v2 mutated_aa_v2 mutation_category_v2  \n",
       "0            AAT             N        nonsynonymous  \n",
       "1            CTT             L           synonymous  \n",
       "2            CTT             L           synonymous  \n",
       "3            CTT             L           synonymous  \n",
       "4            CTT             L           synonymous  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_new['#CHROM'] = \"NC_063383.1\"\n",
    "df1_new = df1_new[['#CHROM', 'site', 'REF_v2', 'ALT', 'nucl_pos_codon', 'parent_codon_v2', 'parent_aa_v2','child_codon_v2','mutated_aa_v2','mutation_category_v2']]\n",
    "df1_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a2f29db7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>mutated_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>186165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM  position REF ALT nucl_pos_codon parent_codon parent_aa  \\\n",
       "0  NC_063383.1    186165   G   A              1          GAT         D   \n",
       "1  NC_063383.1     55133   G   A              3          CTC         L   \n",
       "2  NC_063383.1     55133   G   A              3          CTC         L   \n",
       "3  NC_063383.1     55133   G   A              3          CTC         L   \n",
       "4  NC_063383.1     55133   G   A              3          CTC         L   \n",
       "\n",
       "  child_codon mutated_aa mutation_category  \n",
       "0         AAT          N     nonsynonymous  \n",
       "1         CTT          L        synonymous  \n",
       "2         CTT          L        synonymous  \n",
       "3         CTT          L        synonymous  \n",
       "4         CTT          L        synonymous  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_new.columns = df2.columns\n",
    "df1_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c81e67c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1_new.to_csv(\"article_science_pos_aa.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "f1514dea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>mutated_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>23</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>60</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>78</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>82</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>98</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM  position REF ALT nucl_pos_codon parent_codon parent_aa  \\\n",
       "0  NC_063383.1        23   C   T              -            -         -   \n",
       "1  NC_063383.1        60   C   T              -            -         -   \n",
       "2  NC_063383.1        78   C   T              -            -         -   \n",
       "3  NC_063383.1        82   C   T              -            -         -   \n",
       "4  NC_063383.1        98   C   T              -            -         -   \n",
       "\n",
       "  child_codon mutated_aa mutation_category                  type  \n",
       "0           -          -        intergenic  TC or GA target site  \n",
       "1           -          -        intergenic  TC or GA target site  \n",
       "2           -          -        intergenic  TC or GA target site  \n",
       "3           -          -        intergenic  TC or GA target site  \n",
       "4           -          -        intergenic  TC or GA target site  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2['type'] = 'TC or GA target site'\n",
    "df1_new['type'] = 'Observed APOBEC3-like'\n",
    "df_concat = pd.concat([df2, df1_new], ignore_index=True)\n",
    "df_concat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "65e8c5b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_concat.to_csv(\"combined_table_APOBEC_target.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "49af0807",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>type</th>\n",
       "      <th>number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>Observed APOBEC3-like</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>3608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nonsense</td>\n",
       "      <td>Observed APOBEC3-like</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nonsense</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>Observed APOBEC3-like</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>15162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>synonymous</td>\n",
       "      <td>Observed APOBEC3-like</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>5222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation_category                   type  number\n",
       "0        intergenic  Observed APOBEC3-like     129\n",
       "1        intergenic   TC or GA target site    3608\n",
       "2          nonsense  Observed APOBEC3-like       8\n",
       "3          nonsense   TC or GA target site     703\n",
       "4     nonsynonymous  Observed APOBEC3-like     310\n",
       "5     nonsynonymous   TC or GA target site   15162\n",
       "6        synonymous  Observed APOBEC3-like     184\n",
       "7        synonymous   TC or GA target site    5222"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_grouped = df_concat.groupby(by=[\"mutation_category\", 'type']).size()\n",
    "df_grouped = df_grouped.reset_index()\n",
    "df_grouped = df_grouped.rename(columns={0: 'number'})\n",
    "df_grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2403323d",
   "metadata": {},
   "source": [
    "## grantham_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b9fdf32b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.CHROM</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>nucl_pos_codon</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>mutated_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>type</th>\n",
       "      <th>grantham_score</th>\n",
       "      <th>grantham_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>7584</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TCG</td>\n",
       "      <td>S</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>145</td>\n",
       "      <td>moderately radical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>7595</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>7609</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3.0</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>7624</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ATC</td>\n",
       "      <td>I</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>7629</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>145</td>\n",
       "      <td>moderately radical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>149265</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TCG</td>\n",
       "      <td>S</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>145</td>\n",
       "      <td>moderately radical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8924</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>149281</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>conservative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8925</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>149300</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>GTC</td>\n",
       "      <td>V</td>\n",
       "      <td>GTT</td>\n",
       "      <td>V</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8926</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>149314</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>conservative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8927</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>149359</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTA</td>\n",
       "      <td>L</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>TC or GA target site</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8928 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X.CHROM  position REF ALT  nucl_pos_codon parent_codon parent_aa  \\\n",
       "0     NC_063383.1      7584   C   T             2.0          TCG         S   \n",
       "1     NC_063383.1      7595   C   T             1.0          CTG         L   \n",
       "2     NC_063383.1      7609   C   T             3.0          TTC         F   \n",
       "3     NC_063383.1      7624   C   T             3.0          ATC         I   \n",
       "4     NC_063383.1      7629   C   T             2.0          TCA         S   \n",
       "...           ...       ...  ..  ..             ...          ...       ...   \n",
       "8923  NC_063383.1    149265   G   A             2.0          TCG         S   \n",
       "8924  NC_063383.1    149281   G   A             1.0          CCA         P   \n",
       "8925  NC_063383.1    149300   G   A             3.0          GTC         V   \n",
       "8926  NC_063383.1    149314   G   A             1.0          CAA         Q   \n",
       "8927  NC_063383.1    149359   G   A             1.0          CTA         L   \n",
       "\n",
       "     child_codon mutated_aa mutation_category                  type  \\\n",
       "0            TTG          L     nonsynonymous  TC or GA target site   \n",
       "1            TTG          L        synonymous  TC or GA target site   \n",
       "2            TTT          F        synonymous  TC or GA target site   \n",
       "3            ATT          I        synonymous  TC or GA target site   \n",
       "4            TTA          L     nonsynonymous  TC or GA target site   \n",
       "...          ...        ...               ...                   ...   \n",
       "8923         TTG          L     nonsynonymous  TC or GA target site   \n",
       "8924         TCA          S     nonsynonymous  TC or GA target site   \n",
       "8925         GTT          V        synonymous  TC or GA target site   \n",
       "8926         TAA          *          nonsense  TC or GA target site   \n",
       "8927         TTA          L        synonymous  TC or GA target site   \n",
       "\n",
       "      grantham_score       grantham_rank  \n",
       "0                145  moderately radical  \n",
       "1                  0          synonymous  \n",
       "2                  0          synonymous  \n",
       "3                  0          synonymous  \n",
       "4                145  moderately radical  \n",
       "...              ...                 ...  \n",
       "8923             145  moderately radical  \n",
       "8924               0        conservative  \n",
       "8925               0          synonymous  \n",
       "8926               0        conservative  \n",
       "8927               0          synonymous  \n",
       "\n",
       "[8928 rows x 13 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_score = pd.read_csv(\"grantham_score.csv\", sep=',')\n",
    "df_score_targets = df_score[df_score['type'] == \"TC or GA target site\"]\n",
    "df_score_targets = df_score_targets.reset_index(drop=True)\n",
    "df_score_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f0b42555",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>node</th>\n",
       "      <th>next_x</th>\n",
       "      <th>next_node</th>\n",
       "      <th>grantham_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>parent_aa</td>\n",
       "      <td>S</td>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>L</td>\n",
       "      <td>moderately radical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>L</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>moderately radical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>parent_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>parent_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17851</th>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17852</th>\n",
       "      <td>parent_aa</td>\n",
       "      <td>Q</td>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>*</td>\n",
       "      <td>conservative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17853</th>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>*</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>conservative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17854</th>\n",
       "      <td>parent_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17855</th>\n",
       "      <td>mutated_aa</td>\n",
       "      <td>=</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17856 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                x node      next_x next_node       grantham_rank\n",
       "0       parent_aa    S  mutated_aa         L  moderately radical\n",
       "1      mutated_aa    L        None      None  moderately radical\n",
       "2       parent_aa    =  mutated_aa         =          synonymous\n",
       "3      mutated_aa    =        None      None          synonymous\n",
       "4       parent_aa    =  mutated_aa         =          synonymous\n",
       "...           ...  ...         ...       ...                 ...\n",
       "17851  mutated_aa    =        None      None          synonymous\n",
       "17852   parent_aa    Q  mutated_aa         *        conservative\n",
       "17853  mutated_aa    *        None      None        conservative\n",
       "17854   parent_aa    =  mutated_aa         =          synonymous\n",
       "17855  mutated_aa    =        None      None          synonymous\n",
       "\n",
       "[17856 rows x 5 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#transform table to make ggsankey\n",
    "x = []\n",
    "node = []\n",
    "next_x = []\n",
    "next_node = []\n",
    "grantham_rank_list = []\n",
    "for row in range(0, len(df_score_targets)):\n",
    "    parent_aa = df_score_targets.loc[row, \"parent_aa\"]\n",
    "    mutated_aa = df_score_targets.loc[row, \"mutated_aa\"]\n",
    "    grantham_rank = df_score_targets.loc[row, \"grantham_rank\"]\n",
    "    if parent_aa == mutated_aa:\n",
    "        x.append(\"parent_aa\")\n",
    "        node.append('=')\n",
    "        next_x.append(\"mutated_aa\")\n",
    "        next_node.append('=')\n",
    "        grantham_rank_list.append(grantham_rank)\n",
    "        \n",
    "        x.append(\"mutated_aa\")\n",
    "        node.append('=')\n",
    "        next_x.append(None)\n",
    "        next_node.append(None)\n",
    "        grantham_rank_list.append(grantham_rank)\n",
    "    else:\n",
    "        x.append(\"parent_aa\")\n",
    "        node.append(parent_aa)\n",
    "        next_x.append(\"mutated_aa\")\n",
    "        next_node.append(mutated_aa)\n",
    "        grantham_rank_list.append(grantham_rank)\n",
    "    \n",
    "        x.append(\"mutated_aa\")\n",
    "        node.append(mutated_aa)\n",
    "        next_x.append(None)\n",
    "        next_node.append(None)\n",
    "        grantham_rank_list.append(grantham_rank)\n",
    "    \n",
    "data = {'x': x, 'node': node, \"next_x\":next_x, \"next_node\":next_node, \"grantham_rank\": grantham_rank_list}\n",
    "ggsankey_df = pd.DataFrame.from_dict(data)\n",
    "ggsankey_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a2f44d4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggsankey_df.to_csv(\"ggsankey_targets.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d823cc8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
